### ISSUE TITLE
memory leak with (ora_envhp => 0)

### ISSUE THREAD

**Issue by tomas-zemres** [pythian/DBD-Oracle#27]
There is memory leak if I use connect option: ora_envhp => 0.

Example code:
---
use Memory::Usage;

sub test_connection() {
    my $dbh = DBI->connect( $dsn, $dbuser, '',  { PrintError => 0, ora_envhp => 0 });
    $dbh->disconnect;
 }

my $mu = Memory::Usage->new();
for (1..50) {
    test_connection();
    $mu->record("iteration");
}
$mu->dump();

---

And report says: there is allocated additional 172kB of memory every iteration, but this memory is never released.

If I have process that connects/disconnects to oracle DB (and if it must have set ora_envhp=0, to avoid sharing NLS settings, etc.) then it consumes all available system memory.

This bug is nothing new, but it exists also in older version (for example 1.68).

**Comment by mjegh** [comment]
Which is the most recent version of DBD::Oracle you see this with?

**Comment by tomas-zemres** [comment]
I tested it on current origin/master from this git repository (i.e. commit 18e2d04)

(but this bug is contained also in version 1.68 that I use on production servers).

**Comment by mjegh** [comment]
There does not seem to be anything in dbd_db_disconnect/dbdimp.c to free the environment if ora_envhp was set to false. It should be a simple fix if some state can be saved and tested in dbd_db_disconnect.

**Comment by mjegh** [comment]
or the dbd_db_destroy function which is probably where it should go

**Comment by mjegh** [comment]
IF all your connections use ora_envhp try putting:

```
OCIHandleFree_log_stat(imp_dbh, imp_dbh->envhp, OCI_HTYPE_ENV, status);
```

at the end of dbd_db_destroy in dbdimp.c just after the following line (at the end of that function):

```
OCIHandleFree_log_stat(imp_dbh, imp_dbh->srvhp, OCI_HTYPE_SERVER, status);
```

It is not a proper fix but will determine if the issue is correctly identified.

**Comment by tomas-zemres** [comment]
Yes, simple "Example code" above now releases allocated memory (after this modification of dbdimp.c).

And "Memory::Usage" test report now looks OK.

**Comment by mjegh** [comment]
Thanks, I'll try and write a proper fix for it later today or tomorrow. By all means hassle me if you don't hear anything in a day or two.

**Comment by tomas-zemres** [comment]
I created fix: https://github.com/pythian/DBD-Oracle/pull/28

I hope it is correctly fixed.

**Comment by tomas-zemres** [comment]
I try fix in https://github.com/pythian/DBD-Oracle/pull/28
but it was not correct, because it lead to unexpected exit with status=11.

**Comment by djzort** [comment]
This can probably be closed?
